client/config.c

```diff
@@ -434,3 +457,142 @@ _TDNFConfigReadPluginSettings(
 error:
     goto cleanup;
 }
+
+/*
+ * Read all minimal versions files from pszDir, and store results into
+ * string array pointed to by pppszMinVersions
+ */
+uint32_t
+TDNFReadMinVersionsFiles(
+    char *pszDir,
+    char ***pppszMinVersions
+    )
+{
+    uint32_t dwError = 0;
+    DIR *pDir = NULL;
+    struct dirent *pEnt = NULL;
+    char *pszFile = NULL;
+    char **ppszNewMinVersions = NULL;
+    char ***pppszArrayList = NULL;
+    int nFileCount = 0;
+    int i, j, k;
+    int nLineCount = 0;
+    int nTmp = 0;
+
+    if(IsNullOrEmptyString(pszDir) || !pppszMinVersions)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    /* first, open directory and count all files match *.conf,
+     * so we know how much memory we need */
+    pDir = opendir(pszDir);
+    if (pDir == NULL)
+    {
+        goto cleanup;
+    }
+
+    while((pEnt = readdir(pDir)) != NULL)
+    {
+        if (fnmatch("*.conf", pEnt->d_name, 0) != 0)
+        {
+            continue;
+        }
+        nFileCount++;
+    }
+    closedir(pDir);
+
+    /* allocate memory for our string array */
+    dwError = TDNFAllocateMemory(nFileCount + 1, sizeof(char **), (void **)&pppszArrayList);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    /* read directory again and the files, store content of each file
+     * temporarily in pppszArrayList[i] */
+    i = 0;
+    pDir = opendir(pszDir);
+    while((pEnt = readdir(pDir)) != NULL && i < nFileCount)
+    {
+        if (fnmatch("*.conf", pEnt->d_name, 0) != 0)
+        {
+            continue;
+        }
+        dwError = TDNFAllocateStringPrintf(&pszFile, "%s/%s", pszDir, pEnt->d_name);
+        BAIL_ON_TDNF_ERROR(dwError);
+
+        dwError = TDNFReadFileToStringArray(pszFile, &pppszArrayList[i]);
+        BAIL_ON_TDNF_ERROR(dwError);
+
+        TDNF_SAFE_FREE_MEMORY(pszFile);
+
+        i++;
+    }
+    closedir(pDir);
+    pppszArrayList[i] = *pppszMinVersions;
+
+    /* each file can have multiple lines, count them */
+    for (i = 0; pppszArrayList[i]; i++)
+    {
+        dwError = TDNFStringArrayCount(pppszArrayList[i], &nTmp);
+        BAIL_ON_TDNF_ERROR(dwError);
+
+        nLineCount += nTmp;
+    }
+
+    /* move the lines from 2 dimensional pppszArrayList to
+     * flat pointer list ppszMinVersions */
+    dwError = TDNFAllocateMemory(nLineCount+1, sizeof(char *), (void **)&ppszNewMinVersions);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    for (i = 0, k = 0; pppszArrayList[i]; i++)
+    {
+        for (j = 0; pppszArrayList[i][j]; j++)
+        {
+            ppszNewMinVersions[k++] = pppszArrayList[i][j];
+        }
+        TDNF_SAFE_FREE_MEMORY(pppszArrayList[i]);
+    }
+
+    *pppszMinVersions = ppszNewMinVersions;
```

Comment:
So, here we override pppszMinVersions with the whole list of ppszNewMinVersions?

Reply:
Yes, but we saved the content above (line 531).